sdk/core/System.ClientModel/src/Internal/ArrayBackedPropertyBag.cs

```diff
@@ -0,0 +1,308 @@
+                        rest[_count - 2] = default;
+                        return true;
+                    }
+                }
+
+                return false;
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private bool IsFirst(TKey key) => _first.Key.Equals(key);
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private bool IsSecond(TKey key) => _second.Key.Equals(key);
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void AddInternal(TKey key, TValue value)
+    {
+        switch (_count)
+        {
+            case 0:
+                _first = new Kvp(key, value);
+                _count = 1;
+                return;
+            case 1:
+                if (IsFirst(key))
+                {
+                    _first = new Kvp(_first.Key, value);
+                }
+                else
+                {
+                    _second = new Kvp(key, value);
+                    _count = 2;
+                }
+
+                return;
+            default:
+                if (_rest == null)
+                {
+                    _rest = ArrayPool<Kvp>.Shared.Rent(8);
+                    _rest[_count++ - 2] = new Kvp(key, value);
+                    return;
+                }
+
+                if (_rest.Length <= _count)
+                {
+                    var larger = ArrayPool<Kvp>.Shared.Rent(_rest.Length << 1);
+                    _rest.CopyTo(larger, 0);
+                    var old = _rest;
+                    _rest = larger;
+                    ArrayPool<Kvp>.Shared.Return(old, true);
+                }
+                _rest[_count++ - 2] = new Kvp(key, value);
+                return;
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void SetAt(int index, Kvp value)
+    {
+        if (index == 0)
+            _first = value;
+        else if (index == 1)
+            _second = value;
+        else
+            GetRest()[index - 2] = value;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private TValue GetAt(int index) => index switch
+    {
+        0 => _first.Value,
+        1 => _second.Value,
+        _ => GetRest()[index - 2].Value
+    };
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private int GetIndex(TKey key)
+    {
+        if (_count == 0)
+            return -1;
+        if (_count > 0 && _first.Key.Equals(key))
+            return 0;
+        if (_count > 1 && _second.Key.Equals(key))
+            return 1;
+
+        if (_count <= 2)
+            return -1;
+
+        Kvp[] rest = GetRest();
+        int max = _count - 2;
+        for (var i = 0; i < max; i++)
+        {
+            if (rest[i].Key.Equals(key))
+                return i + 2;
+        }
+        return -1;
+    }
+
+    public void Dispose()
+    {
+#if DEBUG
+        if (_disposed)
+        {
+            return;
+        }
+        _disposed = true;
+#endif
```

Comment:
I wonder why this is done in DEBUG only. The `ArrayPool.Shared.Return` [will throw](https://github.com/dotnet/runtime/blob/72db600a20d581fdc6776edce1863bcf8da0b1cd/src/libraries/System.Private.CoreLib/src/System/Buffers/SharedArrayPool.cs#L124-L127) if we call it twice

Reply:
It looks like this is why: https://github.com/Azure/azure-sdk-for-net/pull/34800
@christothes, do you have a suggestion for how this should change?  If not, I can follow up with a recommendation.

Reply:
I wonder if it is intended as a perf optimization to only branch in debug and assume that covers release cases.  @AlexanderSher, do you know?  Could we always check dispose here to avoid an issue with multiple dispose calls in release builds?

Reply:
Originally, we wanted to have as few fields in the type as possible. Hence the debug-only `_disposed` field. If you look at the implementation, `_rest` is set to `default` during dispose, so the second `Dispose` call won't call `Return` on it, which is checked in [DisposeCreateDispose](https://github.com/Azure/azure-sdk-for-net/pull/34800/files#diff-c3ff41a4de44fe247402f951233d761c0f0f3355c896ac8872976d9c9d678a85R193-R216) test

Reply:
Thanks, @AlexanderSher!  @christothes, let me know if you have further comments on this one, otherwise I'll resolve for now but happy to reopen if needed.